File: test_phase3_complete.py
"""
Phase 3: Timeline Visualization - Comprehensive Testing Suite

This test suite validates all Phase 3 components:
- Timeline Data Processing Service
- Timeline API Endpoints  
- Timeline Visualization Component
- Timeline Network Integration
"""

import os
import requests
import subprocess
import time
from typing import Dict, Any, List

class Phase3ComprehensiveTester:
    def __init__(self):
        self.results = []
        self.backend_url = "http://localhost:8080"
        self.test_user = "phase3_test_user"
        
    def log_result(self, test_name: str, success: bool, details: str = "", data: Any = None):
        """Log test result"""
        result = {
            "test": test_name,
            "success": success,
            "details": details,
            "data": data
        }
        self.results.append(result)
        
        status = "✅" if success else "❌"
        print(f"{status} {test_name}: {details}")
    
    def test_timeline_service_functionality(self):
        """Test timeline data processing service"""
        print(f"\n🔧 Testing Timeline Data Processing Service")
        
        try:
            from services.timeline_service import get_timeline_processor, TimelineProcessor
            
            processor = get_timeline_processor()
            
            if isinstance(processor, TimelineProcessor):
                self.log_result("Timeline Service Import", True, "Service imported successfully")
                
                # Test with comprehensive sample data
                sample_articles = [
                    {
                        'pmid': '12345',
                        'title': 'Machine Learning in Drug Discovery',
                        'authors': ['Smith, J.', 'Doe, A.'],
                        'journal': 'Nature',
                        'year': 2020,
                        'citation_count': 150
                    },
                    {
                        'pmid': '12346',
                        'title': 'AI-Based Cancer Treatment',
                        'authors': ['Johnson, B.', 'Smith, J.'],
                        'journal': 'Science',
                        'year': 2021,
                        'citation_count': 200
                    },
                    {
                        'pmid': '12347',
                        'title': 'Neural Networks for Diagnosis',
                        'authors': ['Brown, C.'],
                        'journal': 'Cell',
                        'year': 2019,
                        'citation_count': 100
                    },
                    {
                        'pmid': '12348',
                        'title': 'Genetic Analysis Methods',
                        'authors': ['Davis, D.', 'Wilson, E.'],
                        'journal': 'Nature Genetics',
                        'year': 2018,
                        'citation_count': 80
                    },
                    {
                        'pmid': '12349',
                        'title': 'Brain Imaging Advances',
                        'authors': ['Taylor, F.'],
                        'journal': 'Neuron',
                        'year': 2022,
                        'citation_count': 50
                    }
                ]
                
                # Test different period strategies
                strategies = ['decade', 'lustrum', 'triennium', 'annual']
                for strategy in strategies:
                    timeline_data = processor.process_articles_for_timeline(
                        sample_articles, 
                        period_strategy=strategy,
                        min_articles_per_period=1
                    )
                    
                    if timeline_data.total_articles == 5:
                        self.log_result(f"Timeline Processing ({strategy})", True, 
                                      f"Processed {timeline_data.total_articles} articles, {len(timeline_data.periods)} periods")
                    else:
                        self.log_result(f"Timeline Processing ({strategy})", False, 
                                      f"Expected 5 articles, got {timeline_data.total_articles}")
                
                # Test milestone identification
                timeline_data = processor.process_articles_for_timeline(sample_articles)
                if len(timeline_data.key_milestones) > 0:
                    self.log_result("Milestone Identification", True, 
                                  f"Identified {len(timeline_data.key_milestones)} milestones")
                else:
                    self.log_result("Milestone Identification", True, "No milestones (expected for test data)")
                
                # Test research evolution analysis
                if isinstance(timeline_data.research_evolution, dict):
                    self.log_result("Research Evolution Analysis", True, 
                                  f"Analyzed {len(timeline_data.research_evolution)} research topics")
                else:
                    self.log_result("Research Evolution Analysis", False, "Research evolution analysis failed")
                    
            else:
                self.log_result("Timeline Service Import", False, "Invalid service instance")
                
        except Exception as e:
            self.log_result("Timeline Service Functionality", False, f"Error: {str(e)}")
    
    def test_timeline_api_endpoints(self):
        """Test timeline API endpoints"""
        print(f"\n🌐 Testing Timeline API Endpoints")
        
        # Test article timeline endpoint
        try:
            response = requests.get(
                f"{self.backend_url}/articles/test001/timeline",
                headers={"User-ID": self.test_user},
                params={"period_strategy": "lustrum", "min_articles": 1},
                timeout=15
            )
            
            if response.status_code == 200:
                data = response.json()
                
                if "timeline_data" in data and "base_article" in data:
                    timeline_data = data["timeline_data"]
                    
                    # Validate response structure
                    required_fields = ["periods", "total_articles", "year_range", "citation_trends", "key_milestones"]
                    missing_fields = [field for field in required_fields if field not in timeline_data]
                    
                    if not missing_fields:
                        self.log_result("Article Timeline API", True, 
                                      f"API working. {timeline_data['total_articles']} articles, {len(timeline_data['periods'])} periods")
                        
                        # Test different period strategies
                        strategies = ['decade', 'lustrum', 'triennium', 'annual']
                        for strategy in strategies:
                            strategy_response = requests.get(
                                f"{self.backend_url}/articles/test001/timeline",
                                headers={"User-ID": self.test_user},
                                params={"period_strategy": strategy, "min_articles": 1},
                                timeout=10
                            )
                            
                            if strategy_response.status_code == 200:
                                strategy_data = strategy_response.json()
                                search_params = strategy_data.get("search_parameters", {})
                                
                                if search_params.get("period_strategy") == strategy:
                                    self.log_result(f"Timeline API ({strategy})", True, 
                                                  f"Strategy {strategy} working")
                                else:
                                    self.log_result(f"Timeline API ({strategy})", False, 
                                                  f"Strategy not applied correctly")
                            else:
                                self.log_result(f"Timeline API ({strategy})", False, 
                                              f"HTTP {strategy_response.status_code}")
                    else:
                        self.log_result("Article Timeline API", False, 
                                      f"Missing fields: {missing_fields}")
                else:
                    self.log_result("Article Timeline API", False, "Invalid response structure")
            else:
                self.log_result("Article Timeline API", False, 
                              f"HTTP {response.status_code}: {response.text[:100]}")
                
        except Exception as e:
            self.log_result("Article Timeline API", False, f"Error: {str(e)}")
        
        # Test project timeline endpoint (expected to fail for test user)
        try:
            response = requests.get(
                f"{self.backend_url}/projects/test-project/timeline",
                headers={"User-ID": self.test_user},
                params={"period_strategy": "lustrum", "min_articles": 2},
                timeout=10
            )
            
            if response.status_code == 404:
                self.log_result("Project Timeline API", True, "Project not found (expected for test user)")
            elif response.status_code == 200:
                self.log_result("Project Timeline API", True, "Project timeline API working")
            else:
                self.log_result("Project Timeline API", False, 
                              f"Unexpected status: {response.status_code}")
                
        except Exception as e:
            self.log_result("Project Timeline API", False, f"Error: {str(e)}")
    
    def test_timeline_component_structure(self):
        """Test timeline visualization component structure"""
        print(f"\n🎨 Testing Timeline Visualization Component")
        
        component_path = "frontend/src/components/TimelineView.tsx"
        
        if not os.path.exists(component_path):
            self.log_result("Timeline Component", False, "TimelineView.tsx not found")
            return
        
        with open(component_path, 'r') as f:
            content = f.read()
        
        # Test component structure
        component_features = [
            ("TypeScript Interfaces", "interface Timeline"),
            ("React Hooks", "useState"),
            ("Effect Hook", "useEffect"),
            ("API Integration", "fetch"),
            ("Period Navigation", "navigatePeriod"),
            ("Article Selection", "onArticleSelect"),
            ("Loading State", "loading"),
            ("Error Handling", "error"),
            ("Period Strategies", "periodStrategy"),
            ("Milestone Display", "showMilestones"),
            ("Timeline Visualization", "periodWidths"),
            ("Responsive Design", "grid-cols"),
            ("Interactive Elements", "onClick"),
            ("Accessibility", "title=")
        ]
        
        for feature_name, feature_code in component_features:
            if feature_code in content:
                self.log_result(f"Component: {feature_name}", True, "Feature implemented")
            else:
                self.log_result(f"Component: {feature_name}", False, "Feature missing")
    
    def test_network_integration(self):
        """Test timeline integration with NetworkView"""
        print(f"\n🔗 Testing Timeline Network Integration")
        
        networkview_path = "frontend/src/components/NetworkView.tsx"
        
        if not os.path.exists(networkview_path):
            self.log_result("Network Integration", False, "NetworkView.tsx not found")
            return
        
        with open(networkview_path, 'r') as f:
            content = f.read()
        
        # Test integration features
        integration_features = [
            ("Timeline Import", "import TimelineView"),
            ("Navigation Mode", "'timeline'"),
            ("Conditional Rendering", "if (navigationMode === 'timeline')"),
            ("Timeline Component Usage", "<TimelineView"),
            ("Props Integration", "pmid={sourceType === 'article'"),
            ("Sidebar Integration", "NetworkSidebar"),
            ("Article Selection", "onArticleSelect={(article)"),
            ("Navigation Button", "Timeline View"),
            ("Color Utility", "getNodeColor"),
            ("Mode Switching", "handleNavigationChange('timeline')")
        ]
        
        for feature_name, feature_code in integration_features:
            if feature_code in content:
                self.log_result(f"Integration: {feature_name}", True, "Feature integrated")
            else:
                self.log_result(f"Integration: {feature_name}", False, "Feature not integrated")
    
    def test_frontend_proxy_routes(self):
        """Test frontend API proxy routes"""
        print(f"\n🔗 Testing Frontend Timeline Proxy Routes")
        
        proxy_routes = [
            ("Article Timeline Proxy", "frontend/src/app/api/proxy/articles/[pmid]/timeline/route.ts"),
            ("Project Timeline Proxy", "frontend/src/app/api/proxy/projects/[id]/timeline/route.ts")
        ]
        
        for route_name, route_path in proxy_routes:
            if os.path.exists(route_path):
                self.log_result(route_name, True, "Proxy route exists")
                
                # Check route content
                with open(route_path, 'r') as f:
                    content = f.read()
                
                route_features = [
                    "NextRequest",
                    "NextResponse", 
                    "GET",
                    "period_strategy",
                    "min_articles",
                    "User-ID"
                ]
                
                missing_features = [f for f in route_features if f not in content]
                
                if not missing_features:
                    self.log_result(f"{route_name} Content", True, "All required features present")
                else:
                    self.log_result(f"{route_name} Content", False, f"Missing: {missing_features}")
            else:
                self.log_result(route_name, False, "Proxy route missing")
    
    def test_phase3_completeness(self):
        """Test Phase 3 overall completeness"""
        print(f"\n🎯 Testing Phase 3 Completeness")
        
        # Define Phase 3 requirements
        phase3_requirements = [
            "Timeline Service Import",
            "Timeline Processing (lustrum)",
            "Article Timeline API",
            "Timeline API (lustrum)",
            "Component: React Hooks",
            "Component: API Integration",
            "Integration: Timeline Import",
            "Integration: Conditional Rendering",
            "Integration: Timeline Component Usage",
            "Article Timeline Proxy"
        ]
        
        completed_requirements = []
        missing_requirements = []
        
        for result in self.results:
            if result["test"] in phase3_requirements:
                if result["success"]:
                    completed_requirements.append(result["test"])
                else:
                    missing_requirements.append(result["test"])
        
        completion_rate = len(completed_requirements) / len(phase3_requirements) * 100
        
        if completion_rate >= 90:
            self.log_result("Phase 3 Completeness", True, 
                          f"Phase 3 {completion_rate:.1f}% complete ({len(completed_requirements)}/{len(phase3_requirements)})")
        else:
            self.log_result("Phase 3 Completeness", False, 
                          f"Phase 3 only {completion_rate:.1f}% complete. Missing: {missing_requirements}")
    
    def run_all_tests(self):
        """Run comprehensive Phase 3 testing"""
        print("🧪 Phase 3: Timeline Visualization - Comprehensive Testing Suite")
        print("=" * 70)
        
        # Test timeline service functionality
        self.test_timeline_service_functionality()
        
        # Test timeline API endpoints
        self.test_timeline_api_endpoints()
        
        # Test timeline component structure
        self.test_timeline_component_structure()
        
        # Test network integration
        self.test_network_integration()
        
        # Test frontend proxy routes
        self.test_frontend_proxy_routes()
        
        # Test overall completeness
        self.test_phase3_completeness()
        
        # Summary
        print("\n📊 Phase 3 Comprehensive Testing Summary")
        print("=" * 50)
        
        total_tests = len(self.results)
        passed_tests = sum(1 for r in self.results if r['success'])
        failed_tests = total_tests - passed_tests
        
        print(f"Total Tests: {total_tests}")
        print(f"✅ Passed: {passed_tests}")
        print(f"❌ Failed: {failed_tests}")
        print(f"Success Rate: {(passed_tests/total_tests)*100:.1f}%")
        
        if failed_tests > 0:
            print("\n❌ Failed Tests:")
            for result in self.results:
                if not result['success']:
                    print(f"  - {result['test']}: {result['details']}")
        
        # Phase 3 status
        if failed_tests == 0:
            status = "✅ PHASE 3 COMPLETE"
            next_step = "Ready for Phase 4: Author-Centric Features & Polish"
        elif failed_tests <= 3:
            status = "⚠️ PHASE 3 MOSTLY COMPLETE"
            next_step = "Minor fixes needed before Phase 4"
        else:
            status = "❌ PHASE 3 NEEDS WORK"
            next_step = "Significant fixes needed"
        
        print(f"\n🎯 Phase 3 Status: {status}")
        print(f"📋 Next Step: {next_step}")
        
        return failed_tests == 0

if __name__ == "__main__":
    tester = Phase3ComprehensiveTester()
    success = tester.run_all_tests()
    
    if success:
        print("\n🚀 Phase 3: Timeline Visualization COMPLETE!")
        print("🎉 Ready to proceed with Phase 4: Author-Centric Features & Polish")
    else:
        print("\n⚠️ Phase 3 has some issues. Please review and fix before proceeding to Phase 4.")

File: test_phase3_complete.js
/**
 * 🚀 PHASE 3 COMPLETE TEST SUITE
 * 
 * Testing all Phase 3 features:
 * 1. Collection Network Features
 * 2. Collaboration Management
 * 
 * Expected: 100% success rate (all tests passing)
 */

const BASE_URL = 'https://r-dagent-production.up.railway.app';
const USER_ID = 'fredericle77@gmail.com';

// Test configuration
const TESTS = [
    // 🔗 Collection Network Features (2 tests)
    {
        name: "Collection Network - Working Collection",
        method: "GET",
        url: `${BASE_URL}/collections/0de0bf09-9ac8-4eed-b632-0261981b204d/network`,
        headers: { "User-ID": USER_ID },
        expectedStatus: 200,
        category: "collection-network"
    },
    {
        name: "Collection Network - Empty Collection",
        method: "GET", 
        url: `${BASE_URL}/collections/34db811e-3d25-4e2f-af01-39c0daf15003/network`,
        headers: { "User-ID": USER_ID },
        expectedStatus: 200,
        category: "collection-network"
    },
    
    // 👥 Collaboration Management (3 tests)
    {
        name: "Collaboration Management Dashboard",
        method: "GET",
        url: `${BASE_URL}/collaborations/manage`,
        headers: { "User-ID": USER_ID },
        expectedStatus: 200,
        category: "collaboration-management"
    },
    {
        name: "Collaboration Invite - Missing Data",
        method: "POST",
        url: `${BASE_URL}/collaborations/invite`,
        headers: { "User-ID": USER_ID, "Content-Type": "application/json" },
        body: JSON.stringify({ project_id: "test" }), // Missing email
        expectedStatus: 400,
        category: "collaboration-management"
    },
    {
        name: "Collaboration Response - Invalid Invitation",
        method: "POST",
        url: `${BASE_URL}/collaborations/respond`,
        headers: { "User-ID": USER_ID, "Content-Type": "application/json" },
        body: JSON.stringify({ invitation_id: "invalid", response: "accept" }),
        expectedStatus: 404,
        category: "collaboration-management"
    }
];

// Test execution
async function runTest(test) {
    const startTime = Date.now();
    
    try {
        console.log(`🧪 [${new Date().toISOString()}] [+${Date.now() - globalStartTime}ms] 🧪 ${test.name}...`);
        
        const options = {
            method: test.method,
            headers: test.headers || {}
        };
        
        if (test.body) {
            options.body = test.body;
        }
        
        const response = await fetch(test.url, options);
        const responseTime = Date.now() - startTime;
        
        // Check status code
        const statusMatch = response.status === test.expectedStatus;
        
        if (statusMatch) {
            console.log(`✅ [${new Date().toISOString()}] [+${Date.now() - globalStartTime}ms] ✅ ${test.name} - SUCCESS`);
            
            // Log response data for successful tests
            if (response.status === 200) {
                try {
                    const data = await response.json();
                    console.log(`VM592:62 Data: ${JSON.stringify({
                        status: response.status,
                        responseTime: `${responseTime}ms`,
                        category: test.category,
                        dataKeys: Object.keys(data),
                        hasData: Object.keys(data).length > 0
                    })}`);
                } catch (e) {
                    console.log(`VM592:62 Data: ${JSON.stringify({
                        status: response.status,
                        responseTime: `${responseTime}ms`,
                        category: test.category,
                        hasData: false
                    })}`);
                }
            } else {
                console.log(`VM592:62 Data: ${JSON.stringify({
                    status: response.status,
                    responseTime: `${responseTime}ms`,
                    category: test.category,
                    expectedError: true
                })}`);
            }
            
            return { success: true, responseTime, status: response.status };
        } else {
            console.log(`❌ [${new Date().toISOString()}] [+${Date.now() - globalStartTime}ms] ❌ ${test.name} - FAILED`);
            console.log(`VM592:64 Expected: ${test.expectedStatus}, Got: ${response.status}`);
            
            // Try to get error details
            try {
                const errorData = await response.text();
                console.log(`VM592:65 Error: ${errorData.substring(0, 200)}...`);
            } catch (e) {
                console.log(`VM592:65 Error: Could not read response body`);
            }
            
            return { success: false, responseTime, status: response.status, expected: test.expectedStatus };
        }
        
    } catch (error) {
        const responseTime = Date.now() - startTime;
        console.log(`❌ [${new Date().toISOString()}] [+${Date.now() - globalStartTime}ms] ❌ ${test.name} - NETWORK ERROR`);
        console.log(`VM592:66 Network Error: ${error.message}`);
        return { success: false, responseTime, error: error.message };
    }
}

// Main execution
const globalStartTime = Date.now();

async function runAllTests() {
    console.log(`🚀 [${new Date().toISOString()}] [+0ms] 🚀 PHASE 3 COMPLETE TEST SUITE STARTING...`);
    console.log(`📊 Total Tests: ${TESTS.length}`);
    console.log(`🎯 Target: 100% Success Rate`);
    console.log('');
    
    const results = [];
    
    for (const test of TESTS) {
        const result = await runTest(test);
        results.push({ ...result, name: test.name, category: test.category });
        
        // Small delay between tests
        await new Promise(resolve => setTimeout(resolve, 100));
    }
    
    // Calculate statistics
    const totalTests = results.length;
    const successfulTests = results.filter(r => r.success).length;
    const failedTests = totalTests - successfulTests;
    const successRate = ((successfulTests / totalTests) * 100).toFixed(1);
    const totalTime = Date.now() - globalStartTime;
    const avgResponseTime = Math.round(results.reduce((sum, r) => sum + r.responseTime, 0) / totalTests);
    
    // Category breakdown
    const categories = {};
    results.forEach(result => {
        if (!categories[result.category]) {
            categories[result.category] = { total: 0, success: 0 };
        }
        categories[result.category].total++;
        if (result.success) categories[result.category].success++;
    });
    
    console.log('');
    console.log('🎉 ===============================================');
    console.log('🎉 PHASE 3 COMPLETE TEST RESULTS');
    console.log('🎉 ===============================================');
    console.log('');
    console.log(`📊 OVERALL PERFORMANCE: ${successRate}% Success Rate`);
    console.log(`⏱️  Total Duration: ${Math.floor(totalTime / 1000 / 60)} minutes ${Math.floor((totalTime / 1000) % 60)} seconds (${totalTime}ms)`);
    console.log(`⚡ Average Response Time: ${avgResponseTime}ms`);
    console.log(`✅ Successful Tests: ${successfulTests}/${totalTests}`);
    console.log(`❌ Failed Tests: ${failedTests}/${totalTests}`);
    console.log('');
    
    // Category breakdown
    console.log('🔍 DETAILED BREAKDOWN BY CATEGORY:');
    console.log('');
    Object.entries(categories).forEach(([category, stats]) => {
        const categoryRate = ((stats.success / stats.total) * 100).toFixed(1);
        const status = categoryRate === '100.0' ? '✅' : categoryRate >= '80.0' ? '⚠️' : '❌';
        console.log(`${status} **${category.toUpperCase()}**: ${categoryRate}% (${stats.success}/${stats.total})`);
    });
    
    console.log('');
    console.log('🎯 PHASE 3 FEATURE STATUS:');
    console.log('');
    console.log(`🔗 **Collection Network Features**: ${categories['collection-network'] ? ((categories['collection-network'].success / categories['collection-network'].total) * 100).toFixed(1) : 0}% Success`);
    console.log(`👥 **Collaboration Management**: ${categories['collaboration-management'] ? ((categories['collaboration-management'].success / categories['collaboration-management'].total) * 100).toFixed(1) : 0}% Success`);
    
    if (successRate === '100.0') {
        console.log('');
        console.log('🎉 🎉 🎉 PHASE 3 COMPLETE - ALL TESTS PASSING! 🎉 🎉 🎉');
        console.log('🚀 Collection Network Features and Collaboration Management are fully operational!');
    } else if (successRate >= '90.0') {
        console.log('');
        console.log('✅ PHASE 3 MOSTLY COMPLETE - Excellent performance!');
        console.log('🔧 Minor issues to address for 100% success rate.');
    } else {
        console.log('');
        console.log('⚠️  PHASE 3 NEEDS ATTENTION - Some features require fixes.');
        console.log('🔧 Review failed tests and implement necessary fixes.');
    }
    
    console.log('');
    console.log('===============================================');
}

// Run the tests
runAllTests().catch(console.error);
